### DISCUSSION TITLE
Why metadata cache generation requires db info at all?

### DISCUSSION THREAD

**Discussion by QuestionAndAnswer** [mikro-orm/mikro-orm#3806] [General]
I'm running mikro-orm cache:generate and without env or in config dbName set, it is failing. This seems like a bug or some bad design choice, but cache:generate should operate only purely with the use of .ts or .d.ts files. For example I can't pregen the cache in a docker, as docker in CI/CD environment will not know any future deployment params.

Moreover, while having a cache and all this metadata things baked makes sense the implementation kind awkward when working with it.
It would be more useful if I can use such caching as usually basic code gen, like tsc is. 
For example:
`mikro-orm cache:generate --out ts,js`, which would generate both ts and js compatible jsons.
But even better I guess there will be no need in such separation in the first place.

As for the development I don't really care about cached results and gonna use TsMorph. For cache generation also TsMorph out of the sources. But for production I need only cache which will be read without need of @mikro-orm/reflection package and ts-morph packages, which are pretty huge for production.

Maybe I didn't get this implementation, but it's hard right now to make a production optimized build as I need to apply small hacks here and there either for cache generation or for testing later. Maybe someone can share how they preparing their builds with the current approach?

**Comment by B4nan** [comment]
This is indeed something I would like to revisit or document better in v6, I know it's far away from ideal if you want to use ts-morph. There are ways, but rather complicated, e.g. using webpack to remove TS and ts-morph from the production bundle (while using the metadata cache for production builds when running). If you check the issues/discussions, I remember one about using a custom metadata provider that would read the cache but throw in case it's missing, that should be rather easy to develop and use (it's an extension point already).

I do have a WIP getting started guide and this is one of the chapters I want to have in there, as the guide is using ts-morph.

**Comment by QuestionAndAnswer** [comment]
@B4nan yes there was a thread: https://github.com/mikro-orm/mikro-orm/discussions/2103

I can share only my setup which is somewhat complicated right now.

I took the solution from thread above, about cache metadata loader and just refactored a bit.

```typescript
import type { EntityMetadata } from '@mikro-orm/core';
import { MetadataProvider, Utils } from '@mikro-orm/core';

// Source: https://github.com/mikro-orm/mikro-orm/discussions/2103

export class CachedMetadataProvider extends MetadataProvider {
    public useCache(): boolean {
        return true;
    }

    // Taken from MikroORM's TsMorphMetadataProvider
    public loadFromCache(meta: EntityMetadata, cache: EntityMetadata): void {
        const values = Object.values(cache.properties);
        for (const prop of values) {
            if (prop.customType) {
                // eslint-disable-next-line no-param-reassign
                prop.customType = meta.properties[prop.name].customType;
            }
        }

        Utils.merge(meta, cache);
    }

    // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
    public async loadEntityMetadata(meta: EntityMetadata, name: string): Promise<void> {}
}

```

When build comes I do the following:

`package.json`
```json
"mikro-orm": {
    "useTsNode": true,
    "configPaths": [
      "./src/orm.cli.ts",
      "./dist/orm.cli.js"
    ]
  }
```
Also, with the setup bellow I was able to exclude @mikro-orm/reflection completely from the runtime, using only cache and cache-metadata-provider. So, package.json contains @mikro-orm/reflection only as a dev-dependency.

I have two configs `orm.ts` and `orm.cli.ts`. I skipped database connection options, as they are not relevant for setup description. But basically I'm just reading manually then from env and substituting (infra reasons mostly).

orm.ts
```typescript
import { CachedMetadataProvider } from './packages/mikro-orm-cache-metadata-provider';

export const baseConfig: MikroOrmOptions = {
    entities: ['./dist/dal/entities'],
    entitiesTs: ['./src/dal/entities'],
    tsNode: false,
    migrations: {
        path: './dist/dal/migrations',
        pathTs: './src/dal/migrations',
        allOrNothing: true,
        transactional: true,
    },
    cache: {
        options: { cacheDir: './dist/dal/mikro-orm-cache' }, // this is the important part, to make cache appear where I need it
        enabled: true,
    },
    type: 'postgresql',
};

export function loadConfig(): MikroOrmOptions {
    return {
        ...baseConfig,
        metadataProvider: CachedMetadataProvider,
    };
}
```

orm.cli.ts
```typescript
import type { MikroOrmOptions } from './orm';
import { loadEnvConfig, baseConfig } from './orm';

/**
 * Default import used by mikro-orm cli or tests calls to read the settings
 * In runtime this would be skipped.
 */
export default async function cliConfig(): Promise<MikroOrmOptions> {
    const mReflection = await import('@mikro-orm/reflection');

    return {
        ...baseConfig,
        metadataProvider: mReflection.TsMorphMetadataProvider,
    };
}
```

This is a base build sequence

1. mikro-orm database:create // That's kind awkward thing to do, can't we have a flag on migration:create? I saw somewhere a ticket for that also.
2. mikro-orm migration:create
3. clean *.ts.json cache // This is the side effect of running previous two commands, and I'm no longer need it
4. tsc // with enabled declaration files
5. mikro-orm cache:generate // This now will cause .js.json cache to appear in dist

At this point, base build have to do 3 discovery runs, and 2 of them would actually generate .json. It's an overkill, but I can't figure out how to make it optimal with current api which mikro-orm provides.

6. For docker additionally, in a docker file I need to set database name env, otherwise mikro-orm will fail to start any command
7. I'm manually removing all **/*.d.ts and **/*.ts.json files. *.ts.json might appear because of the tests here, because they should be based on the sources, not build.

For tests I need to use additional config:

```typescript
import ormCliConfig from '../orm.cli';

async function testConfig(dbName: string): Promise<MikroOrmOptions> {
    const config = await ormCliConfig();

    return {
        ...config,
        tsNode: true, // I need to force tsNode here so it will use .ts.json cache.
        dbName,
        debug: false,
        logger: () => {},
    };
}

```

Additional pain points.

1. I need to manually write migrations up commands for runtime. Maybe a flag at this point would be better?
```typescript
export async function init(config: MikroOrmOptions) {
    const orm = await MikroORM.init<PostgreSqlDriver>(config);

    await orm.getSchemaGenerator().ensureDatabase();

    await orm.getMigrator().up();

    return orm;
}
```
2. Specifying mikro-orm.config file only though env or package, is quite nice, but having a possibility to pass it via cli arg would be less shocking (I mean look at tsc or any other code-gen like tools, all accepts main config files as cli arg).

Idk, but this seems like a really hard way of working with orm.
